src/dbally/llms/clients/local_llm.py

```diff
@@ -0,0 +1,101 @@
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+
+import torch
+from huggingface_hub import login
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from dbally.audit.events import LLMEvent
+from dbally.llms.clients.base import LLMClient, LLMOptions
+from dbally.prompts import ChatFormat
+
+from ..._types import NOT_GIVEN, NotGiven
+
+
+@dataclass
+class LocalLLMOptions(LLMOptions):
+    """
+    Dataclass that represents all available LLM call options for the local LLM client.
+    Each of them is described in the [HuggingFace documentation]
+    (https://huggingface.co/docs/huggingface_hub/en/package_reference/inference_client#huggingface_hub.InferenceClient.text_generation). # pylint: disable=line-too-long
+    """
+
+    repetition_penalty: Union[Optional[float], NotGiven] = NOT_GIVEN
+    do_sample: Union[Optional[bool], NotGiven] = NOT_GIVEN
+    best_of: Union[Optional[int], NotGiven] = NOT_GIVEN
+    max_new_tokens: Union[Optional[int], NotGiven] = NOT_GIVEN
+    top_k: Union[Optional[int], NotGiven] = NOT_GIVEN
+    top_p: Union[Optional[float], NotGiven] = NOT_GIVEN
+    seed: Union[Optional[int], NotGiven] = NOT_GIVEN
+    stop_sequences: Union[Optional[List[str]], NotGiven] = NOT_GIVEN
+    temperature: Union[Optional[float], NotGiven] = NOT_GIVEN
+
+
+class LocalLLMClient(LLMClient[LocalLLMOptions]):
+    """
+    Client for the local LLM that supports Hugging Face models.
+    """
+
+    _options_cls = LocalLLMOptions
+
+    def __init__(
+        self,
+        model_name: str,
+        *,
+        hf_api_key: Optional[str] = None,
+    ) -> None:
+        """
+        Constructs a new local LLMClient instance.
+
+        Args:
+            model_name: Name of the model to use.
+            hf_api_key: The Hugging Face API key for authentication.
+        """
+
+        super().__init__(model_name)
+
+        if hf_api_key:
+            login(hf_api_key)
+
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto", torch_dtype=torch.bfloat16)
```

Comment:
Please use `AutoModel` here. `AutoModelForCausalLM` will raise exception for non `CausalLM` models.